packages/services/src/contents/index.ts

```diff
@@ -1571,3 +1575,259 @@ namespace Private {
     return extension;
   }
 }
+
+class ContentProviderRegistry implements IContentProviderRegistry {
+  constructor() {
+    this._providers = [];
+    this._drive = undefined;
+  }
+
+  get drive(): Contents.IDrive | undefined {
+    return this._drive;
+  }
+
+  set drive(drive: Contents.IDrive) {
+    this._drive = drive;
+    for (let provider of this._providers) {
+      provider.drive = drive;
+    }
+  }
+
+  register(provider: IContentProvider): void {
+    if (this._drive !== undefined) {
+      provider.drive = this._drive;
+    }
+    this._providers.push(provider);
+  }
+
+  getProvider(filePath: string): IContentProvider | undefined {
```

Comment:
What is the use case for matching providers by file extension?

In addition to file extension we _could_ use some of the metadata from `Contents.IFetchOptions`.

Reply:
A provider advertises the type of files it is able to handle through a list of extensions. For instance, an extension could provide access only for HDF5 files.
Yes `Contents.IFetchOptions` could be useful too.